api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -123,11 +160,15 @@ public void checkFreeTierBillingUsage() {
       updateFreeTierWorkspacesStatus(user, BillingStatus.INACTIVE);
     }
 
-    final List<DbUser> allUsers = userDao.findUsers();
-    final Collection<DbUser> usersToThresholdCheck =
-        CollectionUtils.subtract(allUsers, expiredUsers);
+    final Set<DbUser> usersToThresholdCheck =
+        Sets.difference(
+                usersWithChangedCosts,
+                newlyExpiredUsers.stream().map(DbUser::getUserId).collect(Collectors.toSet()))
+            .stream()
+            .map(userDao::findUserByUserId)
```

Comment:
This could also have the same optimization, though we expect n to be small at this point, so it's much less important.